plugins/basic/python/tk_maya_plugin_basic/plugin_logic.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import pymel.core as pm
+
+# For now, import the Shotgun toolkit core included with the plug-in,
+# but also re-import it later to ensure usage of a swapped in version.
+import sgtk as tk_core
+
+import plugin_engine
+
+
+MENU_LOGIN = "ShotgunMenuLogin"
+MENU_LABEL = "Shotgun"
+
+ITEM_LABEL_LOGIN   = "Log In to Shotgun..."
+ITEM_LABEL_LOGOUT  = "Log Out of Shotgun"
+ITEM_LABEL_WEBSITE = "Learn about Shotgun..."
+
+WEBSITE_URL = "https://shotgunsoftware.com"
+
+
+def bootstrap():
+    """
+    Bootstraps the plug-in logic handling user login and logout.
+    """
+
+    if tk_core.authentication.ShotgunAuthenticator().get_default_user():
+        # When the user is already authenticated, automatically log him/her in.
+        login_user()
+    else:
+        # When the user is not yet authenticated, display a login menu.
+        create_login_menu()
+
+
+def shutdown():
+    """
+    Shutdowns the plug-in logic handling user login and logout.
+    """
+
+    if tk_core.platform.current_engine():
+        # When the user is logged in with a running engine, shut down the engine.
+        plugin_engine.shutdown()
+    else:
+        # When the user is logged out, delete the displayed login menu.
+        delete_login_menu()
+
+
+def login_user():
+    """
+    Logs in the user to Shotgun and starts the engine.
+    """
+
+    # Needed global to re-import the toolkit core later.
+    global tk_core
+
+    try:
+        # When the user is not yet authenticated, pop up the Shotgun login dialog to get the user's credentials,
+        # otherwise, get the cached user's credentials.
+        user = tk_core.authentication.ShotgunAuthenticator().get_user()
+
+    except tk_core.authentication.AuthenticationCancelled:
+        # When the user cancelled the Shotgun login dialog, keep around the displayed login menu.
+        return
```

Comment:
Done. I had to use a standalone logger to display messages in Maya script editor when the engine is not running while the user is logged out of Shotgun.